core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -293,27 +331,31 @@ object Queue {
 
         for {
           p <- Promise.make[Nothing, Boolean]
-          _ <- (IO.sync(unsafeOffer(as, p)) *> onQueueEmptySpace(queue) *> p.get)
-                .ensuring(p.poll.void <> IO.sync(unsafeRemove(p)))
+          _ <- (IO.syncSubmit { context =>
+                unsafeOffer(as, p)
+                unsafeOnQueueEmptySpace(queue, context)
+              } *> p.get).ensuring(p.poll.void <> IO.sync(unsafeRemove(p)))
```

Comment:
If `unsafeRemove` is idempotent, you could change `p.poll.void <> IO.sync(unsafeRemove(p))` to `IO.sync(unsafeRemove(p))` and improve performance.

Reply:
`unsafeRemove` will poll the whole queue so it's better not to do it when we can avoid it.

Reply:
Oh but I could do `unsafeRemove` in `onTermination` instead of `ensuring`, then the regular case would be faster.

Reply:
I'm not so sure about that — `ensuring` is a core RTS operation, and is quite fast. But `onTermination` is composite and much slower. This may be ideal. In any case, this is good enough for now, we should do microoptimization separately.